### ISSUE TITLE
EIDM ignores emergencyDecel

### ISSUE THREAD

**Issue by namdre** [eclipse/sumo#9618]
Cars should never break harder than the configured emergencyDecel (which models a physical limit).
If the user wants to see harsh braking, emergencyDecel must be increased explicitly.

EIDM doesn't follow this prescription.

The commend https://github.com/eclipse/sumo/blob/1d017fa2cf2c8b167c78c90b845a07b5d946226d/src/microsim/cfmodels/MSCFModel_EIDM.cpp#L211-L212 suggests some misunderstanding on how emergencyDecel works (or some heretofore unknown bug) and this may have been the reason why it was disabled.

**Comment by namdre** [comment]
@Domsall please comment / fix

**Comment by Domsall** [comment]
The comment is a bit misleading at this point of the Code. There are a few things that influenced the changes in MSCFModel_EIDM::finalizeSpeed:

- From my understanding `vMinEmergency` in `MSCFModel::finalizeSpeed` is the lower bound when decelerating at a predefined stop (do not brake harder than the emergencyDecel at such stop). So it does not bound the overall deceleration at this point?!
- The natural behavior of the IDM is to sometimes brake harder than the predefined decel-value to make a critical situation safe again (unlike the Krauss model). So for example when myDecel is 2.0 and emergencyDecel is 10.0, the Krauss model decelerates with a maximum value of 2.0 and may jump to higher emergency decel values, if needed (disregarding LC-model advice values and patchSpeed).
- The EIDM should model human driving behavior, so I decided to not directly use `patchSpeed`, but rather slowly adopt to `patchSpeed`. I use the parameter `myJerkmax` to limit the jerk and still reach the speed advice from `patchSpeed` at some point.

Are the vehicles braking harder than emergencyDecel? When does this happen?

Do you have a failing / not satisfying test I could use to debug?

I am guessing that this only happens at stops (stopSpeed-calculation), because I recalculate the stopSpeed-velocity in `MSCFModel_EIDM::finalizeSpeed` (see ). I should then maybe bound the decel value here.

**Comment by namdre** [comment]
- vMinEmergency is an absolute lower bound. Whereas the input value to finalize speed 'vPos' may indicate the need for arbitrarily strong braking (i.e. because a leader vehicle happens to brake stronger than anticipated), The permitted braking is limited by emergencyDecel. The speed from processNextStop should generally not be responsible for strong braking since stops aren't "surprising" and cause deceleration well in advance.
- There is no difference between Krauss and IDM with regard to emergencyDecel. As you said, IDM often brakes harder than 'decel' but both models must stay below emergencyDecel. Also, Krauss doesn't necessarily jump to it's full emergencyDecel. The model has code that tries to smooth out deleceration in emergency situations when possible.
- patchSpeed hasn't much to do with emergencyDecel. The value vMin that it gets is influenced by `MAX2(vPos, vMinEmergency)`. Therefore, if vMin indicates strong braking, this is because vPos (carFollowing) requires hard braking.

I've only seen a screenshot today indicating high emergencyDeceleration (-30m/s^s) and noticed that the code doesn't bound deceleration. It should be easy to set up strong braking with a traffic light that switches from green directly to red.

**Comment by Domsall** [comment]
Thanks for the explanations. I will probably just need to add the line from MSCFModel back in (MAX2(vPos, vMinEmergency)). I will do some tests tomorrow.
I tried to create this issue at traffic lights, using the scenarios I already have. But I had problems recreating it -> taking out the yellow phases will help.

About the third point: Yes, you are right. My code documentation is confusing here... I must have deleted the EmergencyDecel-line when I started implementing the model, because I saw weird behavior. However, that behavior did not arise from the EmergencyDecel-line, but rather from the patchSpeed-calculation, where I now use myJerkmax (which solved the problem of fast changing acceleration values in subsecond-simulations).
